R/dominantTaxa.R

```diff
@@ -0,0 +1,210 @@
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' x <- GlobalPatterns
+#'
+#' # Finds the dominant taxa.
+#' x <- dominantTaxa(x)
+#' # Information is stored to colData
+#' colData(x)
+#' #Gets the overview of dominant taxa
+#' overview <- getDominantTaxa(x)
+#' overview
+#'
+#' # If taxonomic information is available, it is possible to find the most dominant
+#' #group from specific taxonomic level, here family level. The name of column can be specified.
+#' x <- dominantTaxa(x, rank="Family", name="dominant_family")
+#' colData(x)
+#' #Gets the overview of dominant taxa
+#' overview <- getDominantTaxa(x)
+#' overview
+#'
+#' x <- microbiomeDataSets::dietswap()
+#' # With group, it is possible to group observations based on group specified by user
+#' x <- dominantTaxa(x, group = "nationality")
+#' #Gets the overview of dominant taxa
+#' overview <- getDominantTaxa(x)
+#' overview
+#'
+NULL
+
+#' @rdname dominantTaxa
+#' @export
+setGeneric("dominantTaxa",signature = c("x"),
+           function(x,
+                    abund_values = "counts",
+                    rank = NULL,
+                    group = NULL,
+                    name = "dominant_taxa")
+               standardGeneric("dominantTaxa"))
+
+
+#' @rdname dominantTaxa
+#' @export
+setMethod("dominantTaxa", signature = c(x = "SummarizedExperiment"),
+          function(x,
+                   abund_values = "counts",
+                   rank = NULL,
+                   group = NULL,
+                   name = "dominant_taxa"){
+
+              # Input check
+              # Check abund_values
+              .check_abund_values(abund_values, x)
+
+              # rank check
+              if(!is.null(rank)){
+                  if(!.is_a_string(rank)){
+                      stop("'rank' must be an single character value.",
+                           call. = FALSE)
+                  }
+                  .check_taxonomic_rank(rank, x)
+              }
+
+              # group check
+              if(!is.null(group)){
+                  if(isFALSE(any(group %in% colnames(colData(x))))){
+                      stop("'group' variable must be in colnames(colData(x))")
+                  }
+              }
+
+              # name check
+              if(!.is_non_empty_string(name)){
+                  stop("'name' must be a non-empty single character value.",
+                       call. = FALSE)
+              }
+
+              # If "rank" is not NULL, species are aggregated according to the taxonomic rank
+              # that is specified by user.
+              if (!is.null(rank)) {
+
+                  # Selects the level
+                  col <- which( taxonomyRanks(x) %in% rank )
+
+                  # Function from taxonomy.R. Divides taxas to groups where they belong.
+                  tax_factors <- .get_tax_groups(x, col = col, onRankOnly = FALSE)
+
+                  # Merges abundances within the groups
+                  mat <- mergeRows(x, f = tax_factors)
+              }
```

Comment:
`mat` is unset if `rank` is NULL.

I think the following would probably solve it:

```
if (!is.null(rank)) {
    # Selects the level
    col <- which( taxonomyRanks(x) %in% rank )

    # Function from taxonomy.R. Divides taxas to groups where they belong.
    tax_factors <- .get_tax_groups(x, col = col, onRankOnly = FALSE)

    # Merges abundances within the groups
    tmp <- mergeRows(x, f = tax_factors)
    mat <- assay(tmp, abund_values)
} else {
    mat <- assay(x, abund_values)
}

```

Reply:
You're correct, mat was missing when rank was NULL. dominantTaxa() also returned object containing modified abundance table. I corrected that. Now the only thing that is modified is the colData where dominant taxa is stored.